### PR TITLE
vk_texture_cache: Implement generic texture cache on Vulkan

### PR DESCRIPTION
**Pull request by ReinUsesLisp**
It currently ignores PBO linearisations since these should be dropped as soon as possible on OpenGL.

### CODE REVIEW

File: src/video_core/texture_cache/surface_params.h [hunk 1]
```diff
@@ -209,6 +209,11 @@ class SurfaceParams {
         return target == VideoCore::Surface::SurfaceTarget::TextureBuffer;
     }
 
+    /// Returns the number of layers in the surface.
+    std::size_t GetNumLayers() const {
+        return is_layered ? depth : 1;
+    }
+
```
Comment (FernandoS27) [inline src/video_core/texture_cache/surface_params.h]: Try not to add Ninja changes into a PR. This should have been in a separate PR.

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+            aspect = vk::ImageAspectFlagBits::eDepth;
+            UNIMPLEMENTED();
+        }
+
+        // Vulkan doesn't seem to understand swizzling of a depth stencil image, use identity
+        swizzle_x = vk::ComponentSwizzle::eR;
+        swizzle_y = vk::ComponentSwizzle::eG;
+        swizzle_z = vk::ComponentSwizzle::eB;
+        swizzle_w = vk::ComponentSwizzle::eA;
+    }
```
Comment (FernandoS27) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp]: add comments here on why is this necessary.

File: src/video_core/texture_cache/surface_params.h (hunk 1 above)
Comment (ReinUsesLisp) [reply src/video_core/texture_cache/surface_params.h]: Moved it to a different commit. It's just moving `GetNumLayers` from private to public and adding a commentary on it.

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp (hunk 2 above)
Comment (ReinUsesLisp) [reply src/video_core/renderer_vulkan/vk_texture_cache.cpp]: Done

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    for (u32 level = 0; level < params.num_levels; ++level) {
+        scheduler.Record([image = image->GetHandle(), buffer = *buffer.handle,
+                          copy = GetBufferImageCopy(level)](auto cmdbuf, auto& dld) {
+            cmdbuf.copyImageToBuffer(image, vk::ImageLayout::eTransferSrcOptimal, buffer, {copy},
+                                     dld);
+        });
+    }
+    scheduler.Finish();
+
+    // TODO(Rodrigo): Use an intern buffer for staging buffers and avoid this unnecesary memcpy.
```
Comment (MysticExile) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp]:
```suggestion
    // TODO(Rodrigo): Use an intern buffer for staging buffers and avoid this unnecessary memcpy.
```
nit

File: src/video_core/renderer_vulkan/vk_texture_cache.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    const vk::BufferView buffer_view;
+    const vk::ImageAspectFlags aspect_mask;
+
+    const VKDevice& device;
+    CachedSurface& surface;
+    const u32 base_layer;
+    const u32 num_layers;
+    const u32 base_level;
+    const u32 num_levels;
+    vk::ImageViewType image_view_type{};
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.h]: This can be const as well, right?

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+                             GPUVAddr gpu_addr, const SurfaceParams& params)
+    : SurfaceBase<View>{gpu_addr, params}, system{system}, device{device},
+      resource_manager{resource_manager}, memory_manager{memory_manager}, scheduler{scheduler},
+      staging_pool{staging_pool} {
+    if (params.IsBuffer()) {
+        buffer = CreateBuffer(device, params);
+        commit = memory_manager.Commit(*buffer, false);
+
+        const auto buffer_view_ci = GenerateBufferViewCreateInfo(device, params, *buffer);
+        format = buffer_view_ci.format;
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp]: Should these be moved to the initializer list, and initialized by a helper function? (that way they can be const)

File: src/video_core/renderer_vulkan/vk_texture_cache.h [hunk 6]
```diff
@@ -0,0 +1,244 @@
+// Copyright 2019 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include <tuple>
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.h]: unused include

File: src/video_core/renderer_vulkan/vk_texture_cache.h [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+#include <unordered_map>
+
+#include <boost/functional/hash.hpp>
+#include <boost/icl/interval_map.hpp>
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.h]: unused includes

File: src/video_core/renderer_vulkan/vk_texture_cache.h [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+#include "video_core/rasterizer_cache.h"
+#include "video_core/renderer_vulkan/declarations.h"
+#include "video_core/renderer_vulkan/vk_image.h"
+#include "video_core/renderer_vulkan/vk_memory_manager.h"
+#include "video_core/renderer_vulkan/vk_scheduler.h"
+#include "video_core/surface.h"
+#include "video_core/texture_cache/surface_base.h"
+#include "video_core/texture_cache/texture_cache.h"
+#include "video_core/textures/decoders.h"
+#include "video_core/textures/texture.h"
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.h]: there are some unused includes in here

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+                      is_linear](auto cmdbuf, auto& dld) {
+        cmdbuf.blitImage(src_image, vk::ImageLayout::eTransferSrcOptimal, dst_image,
+                         vk::ImageLayout::eTransferDstOptimal, {blit},
+                         is_linear ? vk::Filter::eLinear : vk::Filter::eNearest, dld);
+    });
+}
+
+void VKTextureCache::BufferCopy(Surface& src_surface, Surface& dst_surface) {
+    // Currently unimplemented. PBO copies should be dropped and we should use a render pass to
+    // convert from color to depth and viceversa.
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp]: Should we log a warning here or something?

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        {aspect, base_level, num_levels, base_layer, num_layers});
+
+    const auto dev = device.GetLogical();
+    image_view = dev.createImageViewUnique(image_view_ci, nullptr, device.GetDispatchLoader());
+    return last_image_view = *image_view;
+}
+
+bool CachedSurfaceView::IsOverlapping(const View& rhs) const {
+    // TODO(Rodrigo): Also test for layer and mip level overlaps.
+    return &surface == &rhs->surface;
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp]: this is just testing if these are the same surface, right? What about partial overlap?

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+void CachedSurface::DecorateSurfaceName() {
+    // TODO(Rodrigo): Add name decorations
+}
+
+View CachedSurface::CreateView(const ViewParams& params) {
+    return CreateViewInner(params, false);
+}
+
+View CachedSurface::CreateViewInner(const ViewParams& params, bool is_proxy) {
+    auto view = std::make_shared<CachedSurfaceView>(device, *this, params, is_proxy);
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp]: nit: `const` ?

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+    switch (params.target) {
+    case SurfaceTarget::TextureCubemap:
+    case SurfaceTarget::TextureCubeArray:
+        flags |= vk::ImageCreateFlagBits::eCubeCompatible;
+        [[fallthrough]];
+    case SurfaceTarget::Texture1D:
+    case SurfaceTarget::Texture1DArray:
+    case SurfaceTarget::Texture2D:
+    case SurfaceTarget::Texture2DArray:
+        extent = vk::Extent3D(params.width, params.height, 1);
```
Comment (bunnei) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp]: Is 1 right for arrays?

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp (hunk 5 above)
Comment (ReinUsesLisp) [reply src/video_core/renderer_vulkan/vk_texture_cache.cpp]: The problem is that it's initialized in different ways depending on `params.IsBuffer`. Is there a solution for this without repeating code execution at runtime?

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp (hunk 12 above)
Comment (ReinUsesLisp) [reply src/video_core/renderer_vulkan/vk_texture_cache.cpp]: Yes, it means that the depth is 1. On Vulkan number of layers is different to depth.

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp (hunk 10 above)
Comment (ReinUsesLisp) [reply src/video_core/renderer_vulkan/vk_texture_cache.cpp]: I think that's what the `TODO` commentary suggests?

It should test if the view is partially (or fully) overlapping. It's used to determine if a sampled texture is overlapping with the render target, making us change the image layout. Vulkan validation layers will catch if this fails.

I'll rename it to `IsSameSurface` to avoid new callers expecting a fully implemented behaviour.

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp (hunk 5 above)
Comment (ReinUsesLisp) [reply src/video_core/renderer_vulkan/vk_texture_cache.cpp]: Other variables (like `commit`) are also initialized depending on that.
